### DISCUSSION TITLE
StateNotifier as Listenable

### DISCUSSION THREAD

**Discussion by GitGud31** [rrousselGit/riverpod#1013] [Q&A] [answered]
I'm trying out go_router with riverpod, and I was wondering on how to get a Listenable from a riverpod provider. I'm currently using ChangeNotifier and it works fine, but I want to use StateNotifier instead and I don't know how to make it work. Any suggestions on how I can make `refreshListenable` (from go_router) listen to a StateNotifier?

Repo link: [App sample](https://github.com/GitGud31/tst_gorouter_riverpod)

file: root_screen.dart
```
final _router = GoRouter(
...
refreshListenable: ref.watch(authStateNotifier) // I want to use StateNotifier here instead.
...
);
```

file: auth_state_notifier.dart
```
final authStateNotifier =
    ChangeNotifierProvider<AuthStateNotifier>((_) => AuthStateNotifier());

class AuthStateNotifier extends ChangeNotifier {
  bool _isAuthenticated = false;

  bool get isAuthenticated => _isAuthenticated;

  void authenticate() {
    _isAuthenticated = true;
    notifyListeners();
  }

  void logout() {
    _isAuthenticated = false;
    notifyListeners();
  }
}
```

**Comment by iamarnas** [comment, marked as answer]
You handle the simple boolean state. Your `StateNotifier` should look like this:

```dart
class AuthStateNotifier extends StateNotifier<bool>{
  AuthStateNotifier() : super(false);

  void authenticate() {
    state = true;
  }

  void logout() {
    state = false;
  } 
```

**Reply by GitGud31** [reply]
Yes, that's exactly what I did. Thing is the `refreshListenable` (of the go_router) expects a Listenable type which StateNotifier type is not, so I get an error

Here's the code with StateNotifier: 

file: auth_state_notifier.dart
```
import 'package:flutter_riverpod/flutter_riverpod.dart';

final authStateNotifier = StateNotifierProvider<AuthStateNotifier, bool>(
    (_) => AuthStateNotifier(false));

class AuthStateNotifier extends StateNotifier<bool> {
  AuthStateNotifier(state) : super(false);

  bool get isAuthenticated => state;

  void authenticate() {
    state = true;
  }

  void logout() {
    state = false;
  }
}

```

file: root_screen.dart

```
final _router = GoRouter(
   ...
   refreshListenable: ref.watch(authStateNotifier), // error
   ...
);
```

and the Error in that line: 'The argument type 'StateNotifierProvider<AuthStateNotifier, bool>' can't be assigned to the parameter type 'ProviderListenable<Listenable?>'.dartargument_type_not_assignable'

**Reply by iamarnas** [reply]
You can't put `SatateNotifier` to the `Listenable` bouse it does not extend to the `Listenable`. That is why it works with `ChangeNotifier` because it extends to `Listenable`, it is possible `StateNotifier` to convert to the `Stream` but GoRouter required `Listenable` value. If you like `StateNotifier` style you can do it with `ValueNotifier` instead or just use `ChangeNotifier` it's not wrong with it.

```dart
// This will work with Listenable. But needs initialize it over `ref`
final counterProvider = Provider((ref) => Counter());

// This you can use as global everywhere. Works out of the box.
final counterNotifier = Counter();

class Counter extends ValueNotifier<int> {
  Counter() : super(0);
  
  void increment() => value++;
}
```
Or just keep it used with `ChangeNotifier`.

**Reply by iamarnas** [reply]
And tip att the end. You don't need make simple things depend on Riverpod.

> In your case, `ValueNotifier` is easier.

**Reply by GitGud31** [reply]
Yup that's what I noticed (StateNotifier doesn't extend Listenable) checking the doc, so I'll continue to use ChangeNotifier. And I want to ask now, why doesn't StateNotifier extend Listenable like ChangeNotifier does ? as I understand it they're **_almost_** similar (in terms of what they're created for/todo), right?

And thanks for the replies!!

**Reply by GitGud31** [reply]
> In your case, ValueNotifier is easier.

Yes, indeed. But that's just  sample app, so I tried to mimic an auth situation with only a boolean and see how things work (btw riverpod and go_router), Thanks again :D

**Reply by iamarnas** [reply]
> Why doesn't StateNotifier extend Listenable like ChangeNotifier does ?

Because `StateNotifier` not depend on Flutter, it pure Dart. You can move you code to another project to example.

**Reply by GitGud31** [reply]
I see, I appreciate your anwsers, thanks !

**Comment by yeasir-bjit** [comment]
```
@freezed
class AuthState with _$AuthState {
  const AuthState._();
  const factory AuthState.initial() = _Initial;
  const factory AuthState.authenticated() = _Authenticated;
  const factory AuthState.unauthenticated() = _Unauthenticated;
  const factory AuthState.failure(AuthFailure failure) = _Failure;
}
```

My state class looks like this. I am using freezed union for this. How do handle the routing flow based on the authentication state? I my project, oAuth2 is used so that when pressing the login button it opens a webview (modally on iOS). If the user cancels the webview, then authentication fails and the state becomes unauthenticated, and I need to show a failure alert.
